stl/inc/ximmortalize.h

```diff
@@ -0,0 +1,77 @@
+// ximmortalize.h internal header
+
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+#ifndef _XIMMORTALIZE_H
+#define _XIMMORTALIZE_H
+#include <yvals.h>
+#if _STL_COMPILER_PREPROCESSOR
+
+#pragma pack(push, _CRT_PACKING)
+#pragma warning(push, _STL_WARNING_LEVEL)
+#pragma warning(disable : _STL_DISABLED_WARNINGS)
+_STL_DISABLE_CLANG_WARNINGS
+#pragma push_macro("new")
+#undef new
+
+_STD_BEGIN
+#ifdef _M_CEE_PURE
+template <class _Ty>
+union _Immortalizer_impl { // constructs _Ty, never destroys
+    _Immortalizer_impl() noexcept : _Storage{} {}
+    _Immortalizer_impl(const _Immortalizer_impl&) = delete;
+    _Immortalizer_impl& operator=(const _Immortalizer_impl&) = delete;
+    ~_Immortalizer_impl() {
+        // do nothing
+    }
+
+    _Ty _Storage;
+};
+
+#pragma warning(push)
+#pragma warning(disable : 4640) // construction of local static object is not thread-safe (/Wall)
+template <class _Ty>
+constexpr _Ty& _Immortalize() noexcept { // return a reference to an object that will live forever
+    /* MAGIC */ static _Immortalizer_impl<_Ty> _Static;

```

Comment:
I mean, it literally says "magic static" here which is the name of this feature?

Reply:
Thread-safe initialization of local statics is a core language feature; we typically don't have comments that explain how the core language works.

Reply:
Unlike most Core Language features, magic statics are activated near-invisibly with surprising runtime consequences; as  result, when I overhauled the STL to eliminate almost all magic statics, I intentionally and prominently commented the remaining occurrences. I believe that retaining `/* MAGIC */` here, until `/clr:pure` is eradicated, is appropriate.

Reply:
`/clr:pure` isn't the thing that causes strangeness for magic statics; in fact magic statics are 'free' on the CLR.

Reply:
This is in an `_M_CEE_PURE` conditional. Presumably we'll remove the code when `/clr:pure` is eradicated, after which it would be very strange for only `/* MAGIC */` to remain. 😃